include/c2pa.hpp

```diff
@@ -115,4 +125,602 @@ namespace C2pa
+            }
+            add_resource(uri, stream);
+        }    
+
+        /// @brief Add an ingredient to the builder
+        /// @param ingredient_json  Any fields of the ingredient you want to define
+        /// @param format  the format of the ingredient file
+        /// @param source  the input stream to read the ingredient from
+        /// @throws C2pa::Exception for errors encountered by the C2pa library
+        void add_ingredient(const string ingredient_json, const string format, istream& source) {
+            CppIStream c_source = CppIStream(source);
+            int result = c2pa_builder_add_ingredient(builder, ingredient_json.c_str(), format.c_str(), c_source.c_stream);
+            if (result < 0)
+            {
+                throw Exception();
+            }
+        }
+
+        /// @brief Add an ingredient to the builder
+        /// @param ingredient_json  Any fields of the ingredient you want to define   
+        /// @param source_path  the path to the ingredient file
+        /// @throws C2pa::Exception for errors encountered by the C2pa library
+        void add_ingredient(const string ingredient_json, const std::filesystem::path& source_path) {
+            ifstream stream = ifstream(source_path);
+            if (!stream.is_open()) {
+                throw Exception(); // Handle file open error appropriately
+            }
+            auto format = source_path.extension().string();
+            if (!format.empty()) {
+                format = format.substr(1);  // Skip the dot
+            }
+            add_ingredient(ingredient_json, format.c_str(), stream);
+        }    
+
+        /// @brief Sign an input stream and write the signed data to an output stream
+        /// @param format the format of the output stream
+        /// @param source the input stream to sign
+        /// @param dest the output stream to write the signed data to
+        /// @param signer  
+        /// @return a vector containing the signed manifest bytes
+        /// @throws C2pa::Exception for errors encountered by the C2pa library
+        std::vector<const unsigned char>* sign(const string format, istream& source, ostream& dest , Signer& signer) {
+            CppIStream c_source = CppIStream(source);
+            CppOStream c_dest = CppOStream(dest);
+            const unsigned char *c2pa_manifest_bytes = NULL; // todo: make returning manifest bytes optional
+            auto result = c2pa_builder_sign(builder, format.c_str(), c_source.c_stream, c_dest.c_stream, signer.c2pa_signer(), &c2pa_manifest_bytes);
+            if (result < 0)
+            {
+                throw Exception();
+            }
+            if (c2pa_manifest_bytes != NULL) {
+                // Allocate a new vector on the heap and fill it with the data
+                std::vector<const unsigned char>* manifest_bytes = new std::vector<const unsigned char>(c2pa_manifest_bytes, c2pa_manifest_bytes + result);
```

Comment:
Why throw this on the heap? Can we return by value instead? If not, it'd be better to wrap this in a unique_ptr, which is mostly analogous to a `Box`. As long as the unique_ptr is in scope, the underlying data will stay on the heap. When the unque_ptr goes out of scope, the underlying data with be `delete`d. You'd do that like this: 

```suggestion
                    auto c2pa_data = std::make_unique<std::vector<unsigned char>>(c2pa_data_bytes, c2pa_data_bytes + result);
```
And you'd return this: 


```c++
std::unique_ptr<std::vector<unsigned char>>
```